controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -18,66 +18,111 @@ package storageclassclaim
 	"github.com/red-hat-storage/ocs-operator/controllers/storagecluster"
+	providerclient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // StorageClassClaimReconciler reconciles a StorageClassClaim object
 // nolint
 type StorageClassClaimReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
+	Scheme            *runtime.Scheme
+	storageClassClaim *ocsv1alpha1.StorageClassClaim
+	log               logr.Logger
+	storageCluster    *v1.StorageCluster
+	ctx               context.Context
+	namespace         string
 }
 
+const (
+	StorageClassClaimFinalizer = "storageclassclaim.ocs.openshift.io"
+	StorageClassFinalizer      = "storageclass.ocs.openshift.io"
+)
+
 // +kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusters,verbs=get;watch;list
+// +kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
 
 func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx, "StorageClassClaim", request)
-	ctx = ctrllog.IntoContext(ctx, log)
-	log.Info("Reconciling StorageClassClaim.")
+	r.ctx = ctx
+	r.log = ctrllog.Log.WithValues("storageclassclaim", request.NamespacedName)
+	r.log.Info("Reconciling StorageClassClaim.")
+	r.namespace = request.Namespace
 
 	// Fetch the StorageClassClaim instance
-	claim := &ocsv1alpha1.StorageClassClaim{}
-	if err := r.Client.Get(ctx, request.NamespacedName, claim); err != nil {
+	r.storageClassClaim = &ocsv1alpha1.StorageClassClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      request.Name,
+			Namespace: request.Namespace,
+		},
+	}
+	if err := r.get(r.storageClassClaim); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("StorageClassClaim resource not found. Ignoring since object must be deleted.")
+			r.log.Info("StorageClassClaim resource not found. Ignoring since object must be deleted.")
 			return reconcile.Result{}, nil
 		}
-		log.Error(err, "Failed to get StorageClassClaim.")
+		r.log.Error(err, "Failed to get StorageClassClaim.")
 		return reconcile.Result{}, err
 	}
 
```

Comment:
From this point forward we have the claim CR but we do not initialize the `Status.Phase` which means that any failure will just keep the old phase - which I think is incorrect.

Can we set `Status.Phase` to `initializing` here?